### PR TITLE
[Merged by Bors] - Make paused timers update `just_finished` on tick

### LINKED ISSUE #4436: `Timer::just_finished()` returns true forever when paused during finish tick
## Bevy version

0.6.1

## Operating system & version

Ubuntu 20.04

## The problem

The description of `Timer::just_finished()` claims it "returns true _only_ on the tick the timer reached its duration", but pausing _during_ that tick causes it to return true indefinitely. This causes inconsistent behaviour as it means pausing a timer will occasionally make `Timer::just_finished()` return true every frame, usually resulting in a massive speedup for anything the method is being used for: periodic entity spawning, advancing an animation, or printing info to the console.

It's possible this isn't actually a bug, but at the very least it leads to unintuitive behaviour which should be correctly indicated in the method's description.

### Minimal example:

```rust
use bevy::prelude::*;
use std::time::Duration;

fn main() {
    // consistent for both repeating and non-repeating timers
    let mut timer = Timer::from_seconds(1.0, false);

    // finish timer
    timer.tick(Duration::from_secs_f32(1.0));
    assert!(timer.just_finished());
    
    // pause timer while just_finished()
    timer.pause();

    // timer is now just_finished() forever
    for _ in 0..100 {
        timer.tick(Duration::from_secs_f32(0.5));
        assert!(timer.just_finished());
    }
}
```

### What you did

Used a timer to advance an animation when `timer.just_finished() == true`. Paused the timer.

### What you expected to happen

The animation would stop.

### What actually happened

The animation would stop unless it was paused on the tick `timer.just_finished() == true`. If this happened, the animation would speed up to the screen's refresh rate. Unpausing the animation would slow it back down.

## Solutions you have considered
* Make `tick()` reset times_finished to 0 when paused
    * In general, ticking a paused timer is supposed to do nothing.
    * Need to consider what exactly times_finished means while paused / if anything relies on it not changing.

### PR DESCRIPTION
**Pull request by devil-ira**
# Objective
Make timers update `just_finished` on tick, even if paused.
Fixes #4436

## Solution
`just_finished()` returns `times_finished > 0`. So I:
 * Renamed `times_finished` to `times_finished_this_tick` to reduce confusion.
 * Set `times_finished_this_tick` to `0` on tick when paused.
 * Additionally set `finished` to `false` if the timer is repeating.

Notably this change broke none of the existing tests, so I added a couple for this.

Files changed shows a lot of noise because of the rename. Check the first commit for the relevant changes.

## Migration Guide
`Timer::times_finished` has been renamed to `Timer::times_finished_this_tick` for clarity.

### CODE REVIEW

File: crates/bevy_core/src/time/timer.rs [hunk 1]
```diff
@@ -202,6 +202,10 @@ impl Timer {
     /// ```
     pub fn tick(&mut self, delta: Duration) -> &Self {
         if self.paused() {
+            self.times_finished = 0;
```
Comment (alice-i-cecile) [inline crates/bevy_core/src/time/timer.rs:205]: This seems questionable. Presumably times_finished should be usable to get a sense of the total elapsed time.

Why did you need to make this change?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Both of the tests look good, but the fix is a bit questionable to me.

File: crates/bevy_core/src/time/timer.rs (hunk 1 above)
Comment (devil-ira) [reply crates/bevy_core/src/time/timer.rs:205]: Yeah, I was confused about times_finished as well. It's probably not what you think it is. It's not the amount of times a repeating timer has finished in total, but how many times it's finished this tick. I only realized that after reading it's doc comment. Maybe it should be renamed to avoid confusion, and we could probably add a times_finished that actually does what most people would expect it to.

Comment (alice-i-cecile) [reply crates/bevy_core/src/time/timer.rs:205]: > Yeah, I was confused about times_finished as well. It's probably not what you think it is. It's not the amount of times a repeating timer has finished in total, but how many times it's finished within the last tick(i.e. almost always 0 or 1). I only realized that after reading it's doc comment. Maybe it should be renamed to avoid confusion, and we could probably add a times_finished that actually does what most people would expect it to.

I think we should just rename this here: I don't think the functionality is useful enough to add what I had in mind.

**Review by alice-i-cecile — APPROVED** [review]
Aha, that makes way more sense. LGTM now.

**Comment by devil-ira** [comment]
It would be great if this could make it into 0.7.
If we decide we *do* want a `times_finished` that indicates the amount of times a timer has finished since it's last reset for 0.8. We'll end up with a breaking change that still compiles :anguished:

**Comment by alice-i-cecile** [comment]
> It would be great if this could make it into 0.7.

Needs a second approval :) You should consider asking on Discord in #general-dev.

**Comment by hymm** [comment]
I'm not sure about adding this functionality to paused. Real stopwatches don't change the state when they're paused. Only when the reset button is pushed, so holding the finished state makes sense. To me it feels like this should be a separate function that can be called to reset finished.

**Comment by devil-ira** [comment]
The finished state will still be held even with this pr. Only `just_finished` will be updated, unless the the timer is repeating, in which case using `finished()` should behave like `just_finished()` as mentioned in the [example](https://github.com/bevyengine/bevy/blob/193e8c4ada2c324b5222735b0c558398bbbea196/examples/ecs/timers.rs#L59).

Having `just_finished()` return true for more than one tick, (Which currently contradicts it's doc comment!) is IMO suprising, unintuitive and in the usecases I can think of, undesired.
See the linked issue https://github.com/bevyengine/bevy/issues/4436.

**Comment by devil-ira** [comment]
Oh, no. Another victim of the confusingly named variable `times_finished`.
Sorry if that sounded snarky. I think I can speak for @alice-i-cecile and myself that we were very confused at first also. 

I hope you did notice my latest commit renaming `times_finished` to `times_finished_this_tick`. If you're still having trouble, do read the doc comment on the `times_finished_this_tick()` method, and it's tests.

> I think I would prefer if we just decouple the state of just_finished from times_finished. i.e. add a separate bool for just_finished.

When I started making this change this was *exactly* what I was going for, but as I was trying to do so I read the doc comment/tests for `times_finished` and realized that it wasn't what I though it was, and it's very clear to me now that I'm not the only one.

(Pingin @kokounet as they introduced `times_finished`) 

> Also what is the reasoning to setting finished to false if it's paused and repeating?

Purely because it would make [this statement](https://github.com/bevyengine/bevy/blob/193e8c4ada2c324b5222735b0c558398bbbea196/examples/ecs/timers.rs#L59) in the timers example, no longer valid.
I am now leaning more to rephrasing the example and removing the special case I coded in for this.

**Comment by hymm** [comment]
yeah, definitely a bit confusing. So if we pause the timer the internal elapsed time might say evaluate to time_finished_this_tick to 2, but since it's paused it reports 0. But when we unpause the time it'll report 2 or more after the timer is unpaused and ticked, because the internal state hasn't been reset.

> Trying to reason through things.

What we want is for a tick to be a noop to the public api. So is reporting the timer as not having finished correct then? It might be? Honestly not sure. The timer actually finished before you paused the timer...

Really need to see some use cases for this API to know for sure.

**Comment by devil-ira** [comment]
Sorry, I should have explained the purpose of `times_finished_this_tick` better.

Timers do **not** track how many times they have finished, *period*, this has not been added (yet).
Only how many times they finished ***within*** the latest tick.

So, `times_finished_this_tick` can be used to see whether a timer finished multiple times over in one tick.
`finished` and `just_finished` cannot do this.


But I am clearly failing terribly at explaining this with words. I hope these code examples do the trick.

```rust
let mut t = Timer::from_seconds(1.0, true);

if t.tick(Duration::from_secs_f32(3.0)).just_finished() {
    // only reached once!!!
}

for _ in 0..t.tick(Duration::from_secs_f32(3.0)).times_finished_this_tick() {
    // reached three times!!!
}

t.tick(Duration::from_secs_f32(0.0));
assert_eq!(t.times_finished_this_tick(), 0);

t.tick(Duration::from_secs_f32(3.0));
assert_eq!(t.times_finished_this_tick(), 3);

t.tick(Duration::from_secs_f32(69.0));
assert_eq!(t.times_finished_this_tick(), 69);
```

**Comment by kokounet** [comment]
I definitely agree, that name is confusing. At that time I did not think it could be interpreted as the total number of wraps for repeating timers 😅 as I didn't see any use case for it. It's definitely a bad name if it's required to read the doc to understand what it does.
Though I'm not a huge fan of the new name as it is a bit verbose, but still it's better than the current one !

**Comment by hymm** [comment]
I'm sorry to make you explain so much. This was definitely a me problem. I was getting my mental model confused with timers from other programs. Reseting times_finished_this_tick makes sense because this tick is the current/last call to tick, which is a no op if the timer is paused.

**Review by hymm — APPROVED** [review]

**Comment by hymm** [comment]
Should this have a migration guide (label) for the rename?

**Comment by alice-i-cecile** [comment]
@devil-ira please edit your PR description to include the following migration guide section when you get a chance :)

## Migration Guide

`Timer::times_finished` has been renamed to `Timer::times_finished_this_tick` for clarity.
